### PR TITLE
Smac optimizer bugfixes

### PR DESCRIPTION
**Pull request by bpkroth**
- Workaround a bug in SMAC that builds additional_configs across instantiations.
  - [ ] Look for more of these cases (maybe that's where the non-determinism is coming from?)
  - [x] Report it upstream: https://github.com/automl/SMAC3/pull/1067
- Adds more assertions to help check for related issues
- Addresses a TODO sanity check
- Adds some test code for FLAML's surrogate model by increasing iterations
- Updates tests for easier tracing

Note: This still doesn't address all of the non-determinism across platforms/python versions we've noticed recently (#482), and there's some question yet about how much runhistory and some other features are being handled correctly, but this still handles a few small improvements of other issues I've noticed.
For the rest I've filed issues: 
- #487  
- #488

### CODE REVIEW

File: mlos_core/mlos_core/tests/optimizers/optimizer_test.py [hunk 1]
```diff
@@ -197,25 +203,54 @@ def objective(point: pd.DataFrame) -> pd.Series:
...
     assert optimizer is not None
-
     assert llamatune_optimizer is not None
     assert optimizer.optimizer_parameter_space != llamatune_optimizer.optimizer_parameter_space
 
-    for _ in range(num_iters):
+    opt_n_random_init = 0
+    llamatune_n_random_init = 0
+    if 'n_random_init' in kwargs:
+        opt_n_random_init = int(kwargs['n_random_init'])
```
Comment (motus) [inline mlos_core/mlos_core/tests/optimizers/optimizer_test.py:239]: that's just
```suggestion
    llamatune_n_random_init = 0
    opt_n_random_init = int(kwargs.get('n_random_init', 0))
```

File: mlos_core/mlos_core/tests/optimizers/optimizer_test.py [hunk 2]
```diff
@@ -197,25 +203,54 @@ def objective(point: pd.DataFrame) -> pd.Series:
...
+    if optimizer_type == OptimizerType.SMAC:
+        assert isinstance(optimizer, SmacOptimizer)
+        assert isinstance(llamatune_optimizer, SmacOptimizer)
+        # pylint: disable=protected-access
+        opt_n_random_init = optimizer.base_optimizer._initial_design._n_configs
+        llamatune_n_random_init = llamatune_optimizer.base_optimizer._initial_design._n_configs
+
+    for i in range(num_iters):
+        # Place to set a breakpoint for when the optimizer is done with random init.
+        if llamatune_n_random_init and i + 1 >= llamatune_n_random_init:
```
Comment (motus) [inline mlos_core/mlos_core/tests/optimizers/optimizer_test.py:249]: maybe
```suggestion
        if llamatune_n_random_init and i > llamatune_n_random_init:
```

File: mlos_core/mlos_core/tests/optimizers/optimizer_test.py [hunk 3]
```diff
@@ -197,25 +203,54 @@ def objective(point: pd.DataFrame) -> pd.Series:
...
+    opt_n_random_init = 0
+    llamatune_n_random_init = 0
+    if 'n_random_init' in kwargs:
+        opt_n_random_init = int(kwargs['n_random_init'])
+    if optimizer_type == OptimizerType.SMAC:
+        assert isinstance(optimizer, SmacOptimizer)
+        assert isinstance(llamatune_optimizer, SmacOptimizer)
+        # pylint: disable=protected-access
+        opt_n_random_init = optimizer.base_optimizer._initial_design._n_configs
+        llamatune_n_random_init = llamatune_optimizer.base_optimizer._initial_design._n_configs
```
Comment (motus) [inline mlos_core/mlos_core/tests/optimizers/optimizer_test.py:245]: maybe just implement a property `SmacOptimizer.n_configs`? mypy should already know that `optimizer` is of the right type, so we can write `opt_n_random_init = optimizer.n_configs`

File: mlos_core/mlos_core/tests/optimizers/optimizer_test.py [hunk 4]
```diff
@@ -197,25 +203,54 @@ def objective(point: pd.DataFrame) -> pd.Series:
...
+
+    for i in range(num_iters):
+        # Place to set a breakpoint for when the optimizer is done with random init.
+        if llamatune_n_random_init and i + 1 >= llamatune_n_random_init:
+            print("LlamaTuned Optimizer is done with random init.")
+        if opt_n_random_init and i >= opt_n_random_init:
+            print("Optimizer is done with random init.")
```
Comment (motus) [inline mlos_core/mlos_core/tests/optimizers/optimizer_test.py:252]: maybe write to the log instead? we already have the `_LOG` object. (by the way, do we use it anywhere in this file?)

**Review by motus — APPROVED** [review]
Left a few nitpicks

File: mlos_core/mlos_core/tests/optimizers/optimizer_test.py (hunk 4 above)
Comment (bpkroth) [reply mlos_core/mlos_core/tests/optimizers/optimizer_test.py:250]: Good idea.

I added the `_LOG` mostly just to set the level while I was trying to gather information on the non-determinism.

**Pull request merged**